File: src/ast_v2/ast.ts
// Unsupported Rust item: # [derive (Debug , Clone , PartialEq , Eq)] pub enum TypeKind { Struct , Interface , }

// Unsupported Rust item: # [derive (Debug , Clone , PartialEq , Eq)] pub enum TypeRef { Number , String , Bool , Custom (String) , }

// Converted from Rust: struct Field
interface Field {
  name: string;
  ty: TypeRef;
}

// Converted from Rust: struct TypeDecl
interface TypeDecl {
  name: string;
  kind: TypeKind;
  fields: Field[];
}

// Converted from Rust: struct Param
interface Param {
  name: string;
  ty: TypeRef;
}

// Unsupported Rust item: # [derive (Debug , Clone , PartialEq , Eq)] pub enum FunctionKind { Normal , HelloWorldMain , HelloWorldAdd , NeuralNetMain , AstV2ConvertRustFileToTs , AstV2ConvertTsFileToRust , }

// Converted from Rust: struct Function
interface Function {
  name: string;
  params: Param[];
  return_type: TypeRef | undefined;
  kind: FunctionKind;
}

// Converted from Rust: struct Module
interface Module {
  name: string;
  types: TypeDecl[];
  functions: Function[];
}

// Unsupported Rust item: # [allow (dead_code)] pub type TagFn = fn (& mut Module , & Path) ;

File: src/ast_v2/ast.rs
use std::path::Path;

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum TypeKind {
    Struct,
    Interface,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum TypeRef {
    Number,
    String,
    Bool,
    Custom(String),
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Field {
    pub name: String,
    pub ty: TypeRef,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct TypeDecl {
    pub name: String,
    pub kind: TypeKind,
    pub fields: Vec<Field>,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Param {
    pub name: String,
    pub ty: TypeRef,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum FunctionKind {
    Normal,
    HelloWorldMain,
    HelloWorldAdd,
    NeuralNetMain,
    AstV2ConvertRustFileToTs,
    AstV2ConvertTsFileToRust,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Function {
    pub name: String,
    pub params: Vec<Param>,
    pub return_type: Option<TypeRef>,
    pub kind: FunctionKind,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Module {
    pub name: String,
    pub types: Vec<TypeDecl>,
    pub functions: Vec<Function>,
}

// Utility hook for tagging special functions can be reused by
// language-specific adapters if they need the file path.
#[allow(dead_code)]
pub type TagFn = fn(&mut Module, &Path);
